File: array/16.py
from typing import List
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        min_diff = 2 ** 31  - 1
        if len(nums) < 3: return 0
        i = 0
        nums = sorted(nums)
        ret = 0
        while i <= len(nums) - 3:
            if i > 0 and nums[i - 1] == nums[i]:
                i += 1
                continue
            low = i + 1
            high = len(nums) - 1
            base = nums[i]
            while low < high:
                sum_ = base + nums[low] + nums[high]
                if abs(sum_ - target) < min_diff:
                    min_diff = abs(sum_ - target)
                    ret = sum_

                if sum_ < target:
                    low += 1
                    while low < high and nums[low] == nums[low - 1]:
                        low += 1
                elif sum_ == target:
                    return sum_
                else:
                    high -= 1
                    while low < high and nums[high] == nums[high + 1]:
                        high -= 1
            i += 1
        return ret

def main():
    nums = [1, 1, 1, 1]
    target = 0
    print(Solution().threeSumClosest(nums, target))

if __name__ == "__main__":
    main()

File: leetcode-cpp/16.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int min_diff = nums[0] + nums[1] + nums[2];
        sort(nums.begin(), nums.end());

        for (int base = 0; base < nums.size() - 2; base++){
            if (base > 0 && nums[base] == nums[base-1])
                continue;
            int i = base + 1, j = nums.size() - 1, temp;
            while (i < j) {
                temp = nums[base] + nums[i] + nums[j];
                bool l = false, r = false;
                if (abs(min_diff - target) > abs(temp - target))
                    min_diff = temp;
                if (temp == target){
                    return temp;
                } else if (temp < target) {
                    l = true;
                } else{
                    r = true;
                }
                if (l) {
                    i++;
                    while(i < j && nums[i] == nums[i-1]) i++;
                }
                if (r) {
                    j--;
                    while(i < j && nums[j] == nums[j+1]) j--;
                }
            }
        }
        return min_diff;
    }
};

int main(){
    Solution s;
    vector<int> temp({0, 2, 1, -3});
    int target = 1;
    cout << s.threeSumClosest(temp, target) << endl;
//    vector<vector<int>> ret = s.threeSum(temp);
//    for (int i = 0; i < ret.size(); i++){
//        for (int j = 0; j < ret[i].size(); j++){
//            cout << ret[i][j] << "\t";
//        }
//        cout <<endl;
//    }
    return 0;
}
